File: android/app/src/main/java/com/noop/push/ServerRespirationSummary.kt
package com.noop.push

import org.json.JSONObject

/** Selected server respiratory-rate summary; never substitutes a local or vendor scalar. */
data class ServerRespirationSummary(
    val breathsPerMinute: Double?, val mean: Double?, val distribution: List<Double>, val coverage: Double?,
    val acceptedSeconds: Double?, val acceptedWindows: Int?, val totalWindows: Int?, val context: String?,
    val method: String?, val calibrationStatus: String?, val reason: String?, val legacy: Boolean,
    val measurementReason: String? = null,
) {
    companion object {
        fun project(cache: ServerScoreDayCache?, day: String): ServerRespirationSummary? {
            if (cache == null || cache.day != day || cache.ownerId.isBlank()) return null
            val feature = cache.features["respiration"] ?: return null
            val device = feature.deviceId?.takeIf { it.isNotBlank() } ?: return null
            val version = feature.algorithmVersion?.takeIf { it.isNotBlank() } ?: return null
            val overlay = runCatching { JSONObject(cache.rawSnapshotJSON ?: "").getJSONObject("server_scoring") }.getOrNull() ?: return null
            val rawFeature = overlay.optJSONObject("features")?.optJSONObject("respiration") ?: return null
            if (!overlay.optString("user_id").equals(cache.ownerId, ignoreCase = true) || overlay.optString("day") != day ||
                rawFeature.optString("device_id") != device || rawFeature.optString("algorithm_version") != version) return null
            fun number(value: Any?): Double? = (value as? Number)?.toDouble()?.takeIf { it.isFinite() }
            fun count(value: Any?): Int? = number(value)?.takeIf { it >= 0 && it <= Int.MAX_VALUE && it % 1.0 == 0.0 }?.toInt()
            fun text(obj: JSONObject, key: String): String? = (obj.opt(key) as? String)?.takeIf { it.isNotBlank() }
            val daily = overlay.optJSONObject("daily") ?: JSONObject()
            val legacy = version == "frwhoop-server-1"
            val summary = daily.optJSONObject("respiration_summary") ?: JSONObject()
            val scalar = number(daily.opt("resp_rate_bpm"))?.takeIf { it > 0 }
            val median = number(summary.opt("median_bpm"))?.takeIf { it > 0 }
            val mean = number(summary.opt("mean_bpm"))?.takeIf { it > 0 }
            val coverage = number(summary.opt("coverage"))?.takeIf { it in 0.0..1.0 }
            val seconds = number(summary.opt("accepted_seconds"))?.takeIf { it >= 0 }
            val accepted = count(summary.opt("accepted_windows"))
            val total = count(summary.opt("total_windows"))
            val context = text(summary, "context")
            val values = summary.optJSONArray("distribution_bpm")
            val distribution = (0 until (values?.length() ?: 0)).mapNotNull { number(values?.opt(it))?.takeIf { n -> n > 0 } }.sorted()
            val visible = feature.status in setOf("available", "fresh", "stale")
            val measurementReason = text(daily, "respiration_unavailable_reason")
            var reason = feature.reason ?: measurementReason
            var primary = if (visible) scalar else null
            if (!visible) reason = reason ?: "respiration_unavailable"
            if (legacy) {
                reason = reason ?: "legacy_quality_unavailable"
            } else if (primary != null) {
                if (context != "main_sleep") {
                    primary = null; reason = "incompatible_respiration_context"
                } else if (median != scalar || mean == null || coverage == null || coverage == 0.0 || seconds == null || seconds == 0.0 ||
                    accepted == null || accepted == 0 || total == null || accepted > total) {
                    primary = null; reason = "inconsistent_respiration_summary"
                }
            }
            if (primary == null) reason = reason ?: "no_quality_eligible_windows"
            return ServerRespirationSummary(primary, if (legacy) null else mean, if (legacy) emptyList() else distribution,
                if (legacy) null else coverage, if (legacy) null else seconds, if (legacy) null else accepted,
                if (legacy) null else total, if (legacy) null else context, if (legacy) null else text(summary, "method_version"),
                if (legacy) null else text(summary, "calibration_status"), reason, legacy,
                if (primary == null) measurementReason else null)
        }
    }
}

File: Packages/WhoopStore/Sources/WhoopStore/ServerRespirationSummary.swift
import Foundation
import CoreFoundation

/// A selected server respiratory-rate summary, never a local or vendor fallback.
public struct ServerRespirationSummary: Equatable {
    public let breathsPerMinute: Double?
    public let mean: Double?
    public let distribution: [Double]
    public let coverage: Double?
    public let acceptedSeconds: Double?
    public let acceptedWindows: Int?
    public let totalWindows: Int?
    public let context: String?
    public let method: String?
    public let calibrationStatus: String?
    public let reason: String?
    /// Measurement eligibility is separate from snapshot freshness (for example newer_input_pending).
    public let measurementReason: String?
    public let legacy: Bool

    public static func project(_ cache: ServerScoreDayCache?, day: String) -> Self? {
        guard let cache, cache.day == day, !cache.ownerId.isEmpty,
              let feature = cache.features["respiration"],
              let device = feature.deviceId, !device.isEmpty,
              let version = feature.algorithmVersion, !version.isEmpty,
              let bytes = cache.rawSnapshotJSON?.data(using: .utf8),
              let root = (try? JSONSerialization.jsonObject(with: bytes)) as? [String: Any],
              let overlay = root["server_scoring"] as? [String: Any],
              (overlay["user_id"] as? String)?.lowercased() == cache.ownerId.lowercased(),
              overlay["day"] as? String == day,
              let features = overlay["features"] as? [String: [String: Any]],
              features["respiration"]?["device_id"] as? String == device,
              features["respiration"]?["algorithm_version"] as? String == version else { return nil }
        func number(_ value: Any?) -> Double? {
            guard let n = value as? NSNumber, CFGetTypeID(n) != CFBooleanGetTypeID(), n.doubleValue.isFinite else { return nil }
            return n.doubleValue
        }
        func nonnegativeInt(_ value: Any?) -> Int? {
            guard let n = number(value), n >= 0, n < Double(Int.max), n.rounded() == n else { return nil }
            return Int(n)
        }
        let daily = overlay["daily"] as? [String: Any] ?? [:]
        let legacy = version == "frwhoop-server-1"
        let summary = daily["respiration_summary"] as? [String: Any] ?? [:]
        let scalar = number(daily["resp_rate_bpm"]).flatMap { $0 > 0 ? $0 : nil }
        let median = number(summary["median_bpm"]).flatMap { $0 > 0 ? $0 : nil }
        let mean = number(summary["mean_bpm"]).flatMap { $0 > 0 ? $0 : nil }
        let coverage = number(summary["coverage"]).flatMap { (0...1).contains($0) ? $0 : nil }
        let seconds = number(summary["accepted_seconds"]).flatMap { $0 >= 0 ? $0 : nil }
        let accepted = nonnegativeInt(summary["accepted_windows"])
        let total = nonnegativeInt(summary["total_windows"])
        let context = summary["context"] as? String
        let distribution = (summary["distribution_bpm"] as? [Any] ?? []).compactMap(number).filter { $0 > 0 }.sorted()
        let visible = ["available", "fresh", "stale"].contains(feature.status)
        let measurementReason = daily["respiration_unavailable_reason"] as? String
        var reason = feature.reason ?? measurementReason
        var primary = visible ? scalar : nil
        if !visible { reason = reason ?? "respiration_unavailable" }
        if legacy {
            reason = reason ?? "legacy_quality_unavailable"
        } else if primary != nil {
            if context != "main_sleep" {
                primary = nil; reason = "incompatible_respiration_context"
            } else if median != scalar || mean == nil || coverage == nil || coverage == 0 || seconds == nil || seconds == 0 ||
                        accepted == nil || accepted == 0 || total == nil || accepted! > total! {
                primary = nil; reason = "inconsistent_respiration_summary"
            }
        }
        if primary == nil { reason = reason ?? "no_quality_eligible_windows" }
        return Self(breathsPerMinute: primary, mean: legacy ? nil : mean, distribution: legacy ? [] : distribution,
                    coverage: legacy ? nil : coverage, acceptedSeconds: legacy ? nil : seconds,
                    acceptedWindows: legacy ? nil : accepted, totalWindows: legacy ? nil : total,
                    context: legacy ? nil : context, method: legacy ? nil : summary["method_version"] as? String,
                    calibrationStatus: legacy ? nil : summary["calibration_status"] as? String,
                    reason: reason, measurementReason: primary == nil ? measurementReason : nil, legacy: legacy)
    }
}
